pkg/code-generator/schemagen/README.md

```diff
@@ -0,0 +1,49 @@
+# Schemagen
+A tool used to generate validation schemas for Kubernetes CRDs
+
+## Validation Schemas
+Custom Resources in Kubernetes support [defining a structural schema](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema), which is validated during creation and updates. V1 CRDs (required in Kube 1.22 and beyond) require validation schemas. These schemas provide syntax validation (ie does the configuration respect the API structure).
+
+## Options
+`crdDirectory` - Path to the directory where CRDs will be read from and written to
+
+`jsonSchemaTool` - Name of the tool used to generate JsonSchemas, defaults to `protoc`
+
+`removeDescriptionsFromSchema` - Whether to remove descriptions from validation schemas, defaults to `false`
+
+`enumAsIntOrString` - Whether to assign Enum fields the `x-kubernetes-int-or-string` property which allows the value to either be an integer, or a string, defaults to `false`. Only strings are supported when this is false.
+
+`MessagesWithEmptySchema` - A list of message names (ie `core.solo.io.Status`) for which we should generate a validation schema that accepts all properties. This is useful when certain messages take too long to compute schemas, or are recursive.
+
+## Implementation
+This tool executes the following steps to generate validation schemas:
+1. Determine if a project is configured to generate validation schemas. If not, return immediately.
+2. Read all the CRDs from the provided `crdDirectory`
+3. Choose the `JsonSchemaTool`. We currently support `cue` and `protoc` though `protoc` is used in Gloo Edge
+4. Generate all JsonSchemas for the project
+5. For each CRD
+   - Match the CRD with the JsonSchema, using the GVK.
+   - Write the modified CRD back to the file, using the GVK to generate the file name
+
+**NOTE** It would be nice to move complete CRD generation into this tool, but for now we only modify the schema.
+
+## Generation Tools
+We currently support 2 separate implementations for schema generation.
+
+### [Cue](github.com/solo-io/cue)
+`Use cuelang as an intermediate language for transpiling protobuf schemas to openapi v3 with k8s structural schema constraints.`
+
+This is our preferred implementation. It is not used in production yet, due to some performance issues with the Gloo Edge API. However, our goal is to eventually migrate our code to rely on this implementation. I added in this code in this iteration so that we could compare the generated schemas between the implementations.
```

Comment:
```suggestion
This is our preferred implementation. It is not used in production yet, due to some performance issues with the Gloo Edge API. However, our goal is to eventually migrate our code to rely on this implementation. This code was included in this iteration so that we could compare the generated schemas between the implementations.
```